x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/stats/StatsDirectoryWrapper.java

```diff
@@ -0,0 +1,260 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+package org.elasticsearch.index.store.stats;
+
+import org.apache.lucene.store.Directory;
+import org.apache.lucene.store.FilterDirectory;
+import org.apache.lucene.store.IOContext;
+import org.apache.lucene.store.IndexInput;
+import org.elasticsearch.common.util.concurrent.ConcurrentCollections;
+
+import java.io.IOException;
+import java.util.Map;
+import java.util.Objects;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.concurrent.atomic.LongAdder;
+import java.util.function.LongConsumer;
+import java.util.stream.Collectors;
+
+/**
+ * Directory wrapper that records various information about how {@link IndexInput} are accessed.
+ */
+public class StatsDirectoryWrapper extends FilterDirectory {
+
+    private final Map<String, LiveStats> records = ConcurrentCollections.newConcurrentMap();
+
+    protected StatsDirectoryWrapper(final Directory in) {
+        super(in);
+    }
+
+    public Map<String, IndexInputStats> getStats() {
+        return records.entrySet().stream().collect(Collectors.toMap(Map.Entry::getKey, e -> e.getValue().toIndexInputStats()));
+    }
+
+    public IndexInputStats getStatsOrNull(final String name) {
+        final LiveStats stats = records.get(name);
+        if (stats != null) {
+            return stats.toIndexInputStats();
+        }
+        return null;
+    }
+
+    @Override
+    public IndexInput openInput(final String name, final IOContext context) throws IOException {
+        final LiveStats stats = records.computeIfAbsent(name, n -> new LiveStats()).incrementOpenCount();
+        return new StatsIndexInputWrapper(super.openInput(name, context), stats);
+    }
+
+    @Override
+    public void close() throws IOException {
+        try {
+            super.close();
+        } finally {
+            records.clear();
+        }
+    }
+
+    /**
+     * Wraps an {@link IndexInput} to record stats about how it is accessed.
+     */
+    private static class StatsIndexInputWrapper extends IndexInput {
+
+        /**
+         * The IndexInput to delegate operations to
+         **/
+        private final IndexInput input;
+        /**
+         * The statistics to update on IndexInput operations
+         **/
+        private final LiveStats stats;
+        /**
+         * The last read position is kept around in order to detect (non)contiguous reads
+         **/
+        private final AtomicLong lastReadPosition;
```

Comment:
Why do you need an AtomicLong ?

Reply:
It is not needed, I thought I removed it already but :man_facepalming: